Tutorial/jsonscheme (#114)
gardener/landscaper: docs/tutorials/05-external-jsonschema.md

- _definitions_: component that contains the jsonschema definition
- _echo-server_: simple echo server that consumes the jsonschema.

#### Resources JSONSchema

First the shared jsonschema has to be created.
The Kubernetes resource definition is here used as an example.

```
# ./docs/tutorials/resources/external-jsonschema/definitions/component-descriptor
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "landscaper.gardener.cloud/tutorial/external-jsonschema/resources",
  "title": "Resources",
  "description": "Describes kubernetes resource requests and limits",
  "type": "object",
  "properties": {
    "requests": {
      "$ref": "#/definitions/resource"
    },
    "limits": {
      "$ref": "#/definitions/resource"
    }
  },
  "definitions": {
    "resource": {
      "properties": {
        "cpu": {
          "type": "string"
        },
        "memory": {
          "type": "string"
        }
      }
    }
  }
}
```

When the jsonschema is defined it is added to the component descriptor as local artifact to the component descriptor.
For more detailed explanation have a look at the [second tutorial](./02-local-simple-blueprint.md)
   

> Therefore, create the resource definition, create the component descriptor, add the resource with the component-cli and upload the component descriptor.

Reviewer: ```suggestion
Now, create the resource definition and component descriptor, add the resource with the component-cli and upload the component descriptor.
```

The result is a component descriptor that contains the previously created jsonschema as local resource.

<details>

```yaml
# ./docs/tutorials/resources/external-jsonschema/definitions/jsonschema-resource.yaml
---
type: landscaper.gardener.cloud/jsonschema
name: resources-definition
relation: local
input:
  type: "file"
  path: "./resources.json"
  mediaType: "application/vnd.gardener.landscaper.jsonscheme.v1+json"
...
```
```yaml
# ./docs/tutorials/resources/external-jsonschema/definitions/component-descriptor.yaml
meta:
  schemaVersion: v2
component:
  name: github.com/gardener/landscaper/external-jsonschema/definitions
  provider: internal
  repositoryContexts:
  - baseUrl: eu.gcr.io/gardener-project/landscaper/tutorials/components
    type: ociRegistry
  resources: []
  componentReferences: []
  sources: []
```

```
component-cli ca resources add ./docs/tutorials/resources/external-jsonschema/definitions -r ./docs/tutorials/resources/external-jsonschema/definitions/jsonscheme-resource.yaml -v 5
```

```
component-cli ca remote push ./docs/tutorials/resources/external-jsonschema/definitions
```

</details>

#### Echo Server Deployment

> The first component is now created and uploaded to the registry.

Reviewer: ```suggestion
With the _definitions_  component ready, let's move on and create a component which will consume those definitions.```

> Now the user of the jsonschema is created.

Reviewer: ```suggestion
```

> The echo server example is reused and enhanced with another import parameter _resources_.

Reviewer: ```suggestion
To keep things as simple as possible, the echo server example will be reused and enhanced with another import parameter _resources_.
```

Before the jsonschema can be used in the blueprint, it has to be referenced in the component descriptor so that Landscaper knows about the dependency towards the `definitions` component.

> Subsequently, a component reference is added (see the complete component descriptor in the details):

Reviewer: ```suggestion
Hence, a component reference entry is added to the echo server's component descriptor (see the complete component descriptor in the details):
```

```yaml
componentReferences:
- name: definitions
componentName: github.com/gardener/landscaper/external-jsonschema/definitions
version: v0.1.0
```

<details>
<div id="echo-server-comp-desc"></div>

```yaml
# ./docs/tutorials/resources/external-jsonschema/echo-server/component-descriptor.yaml
meta:
  schemaVersion: v2

component:
name: github.com/gardener/landscaper/external-jsonschema/echo-server
version: v0.1.0

provider: internal

repositoryContexts:
- type: ociRegistry
  baseUrl: eu.gcr.io/gardener-project/landscaper/tutorials/components

sources: []
componentReferences:
- name: definitions
  componentName: github.com/gardener/landscaper/external-jsonschema/definitions
  version: v0.1.0

resources:
- type: ociImage
  name: echo-server-image
  version: v0.2.3
  relation: external
  access:
  type: ociRegistry
  imageReference: hashicorp/http-echo:0.2.3
```

</details>

> When the component reference is added to the echo-server's component descriptor, the resource import can be added to the blueprint.

Reviewer: ```suggestion
Based on the extended component descriptor, the resource import can be added to the blueprint.
```

The import also contains a valid jsonschema but uses the `$ref` attribute with a custom landscaper protocol implementation.
That custom protocol specifies that the jsonschema should be fetched from the resource `resources-definition` of the referenced component `definitions`.

For detailed explanation about available reference methods see the [blueprint jsonschema documentation](../usage/JSONSchema.md)

```yaml
imports:
- name: resources
  schema:
    $ref: "cd://componentReferences/definitions/resources/resources-definition"
```

The blueprint is defined, so it can be added to the component descriptor using the component-cli and then uploaded.
See the details section for concrete resource definitions and cli calls.

<details>

```yaml
# ./docs/tutorials/resources/external-jsonschema/echo-server/blueprint.yaml
apiVersion: landscaper.gardener.cloud/v1alpha1
kind: Blueprint

imports:
- name: cluster
  targetType: landscaper.gardener.cloud/kubernetes-cluster
- name: ingressClass
  schema:
    type: string
- name: resources
  schema:
    $ref: "cd://componentReferences/definitions/resources/resources-definition"

deployExecutions:
- name: default
  type: GoTemplate

>   file: /defaultDeployExecution.yaml

Reviewer: should the DeployExecution which consumes the resources import be listed here too?

Author: oh yes good point

```

```yaml
# ./docs/tutorials/resources/external-jsonschema/echo-server/blueprint-resource.yaml
---
type: blueprint
name: echo-server-blueprint
version: v0.1.0
relation: local
input:
  type: "dir"
  path: "./blueprint"
  compress: true
  mediaType: "application/vnd.gardener.landscaper.blueprint.v1+tar+gzip"
...
```

```
component-cli ca resources add ./docs/tutorials/resources/external-jsonschema/echo-server -r ./docs/tutorials/resources/external-jsonschema/echo-server/blueprint-resource.yaml -v 5
```

```
component-cli ca remote push ./docs/tutorials/resources/external-jsonschema/echo-server
```

</details>

### Installation

The components with their jsonschema and blueprint artifacts are now uploaded and can be deployed.

> To deploy the component using Landscaper, first the target import and the data imports have to be defined and applied to the Kubernetes cluster.

Reviewer: ```suggestion
In a first step, the target import and the data imports have to be defined and applied to the Kubernetes cluster so that Landscaper can pick them up.
```

```yaml
# ./docs/tutorials/resources/external-jsonschema/my-target.yaml
apiVersion: landscaper.gardener.cloud/v1alpha1
kind: Target
metadata:
  name: my-target-cluster
spec:
  type: landscaper.gardener.cloud/kubernetes-cluster
  config:
    kubeconfig: |
      apiVersion:...
      # here goes the kubeconfig of the target cluster
```

```yaml
# ./docs/tutorials/resources/external-jsonschema/configmap.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-imports
data:
  ingressClass: "nginx"
  resources: |
    requests:
      memory: 50Mi
    limits:
      memory: 100Mi
```

```
kubectl apply -f ./docs/tutorials/resources/external-jsonschema/my-target.yaml
kubectl apply -f ./docs/tutorials/resources/external-jsonschema/configmap.yaml
```

> The imports are now available in the system, so the installation can be applied and will be installed.

Reviewer: ```suggestion
The imports are now available in the system, so the installation can be applied and will be processed by Landscaper.
```
